src/range_query/sqrt_decomp.rs

```diff
@@ -129,14 +131,54 @@ impl PiecewiseLinearFn {
     fn update_envelope(&mut self) {
         self.recent_lines.extend(self.sorted_lines.drain(..));
         self.recent_lines.sort_unstable();
-        for (slope, intercept) in self.recent_lines.drain(..) {
-            // TODO: do convex hull trick algorithm
-            self.sorted_lines.push((slope, intercept));
+        self.intersections.clear();
+
+        'outer: for (m1, b1) in self.recent_lines.drain(..) {
+            while let Some(&(m2, b2)) = self.sorted_lines.last() {
+                // The lines are sorted by intercept so an incoming line with
+                // the same slope can never beat the previous line.
+                if m2 == m1 {
+                    continue 'outer;
+                }
+                let new_intersection = (b1 - b2) as f64 / (m2 - m1) as f64;
+                if let Some(&last_intersection) = self.intersections.last() {
+                    if new_intersection < last_intersection {
```

Comment:
Flipping this condition to `if new_intersection > self.intersections.last().unwrap_or(f64::MIN)` (and breaking if so) would eliminate all uses of `continue`. This change also prevents inserting the same intersection point multiple times (because we pop in the `==` case).

Reply:
Great! I thought of flipping it but couldn't figure out a way past the possibly-missing last intersection. My brain was stuck on the "let Some(...)" construction. I did think of adding a dummy -INF intersection to the array in the beginning, but this way looks better.